contracts/peg/NonCustodialPSM.sol

```diff
@@ -0,0 +1,452 @@
+        amountFeiOut = _getMintAmountOut(amountIn);
+    }
+
+    /// @notice calculate the amount of underlying out for a given `amountFeiIn` of FEI
+    /// First get oracle price of token
+    /// Then figure out how many dollars that amount in is worth by multiplying price * amount.
+    /// ensure decimals are normalized if on underlying they are not 18
+    function getRedeemAmountOut(uint256 amountFeiIn)
+        public
+        view
+        override
+        returns (uint256 amountTokenOut)
+    {
+        amountTokenOut = _getRedeemAmountOut(amountFeiIn);
+    }
+
+    /// @notice the maximum mint amount out
+    function getMaxMintAmountOut() external view override returns (uint256) {
+        return fei().balanceOf(address(this)) + buffer();
+    }
+
+    // ----------- Internal Methods -----------
+
+    /// @notice helper function to get mint amount out based on current market prices
+    /// @dev will revert if price is outside of bounds and bounded PSM is being used
+    function _getMintAmountOut(uint256 amountIn)
+        internal
+        view
+        virtual
+        returns (uint256 amountFeiOut)
+    {
+        Decimal.D256 memory price = readOracle();
+        _validatePriceRange(price);
+
+        Decimal.D256 memory adjustedAmountIn = price.mul(amountIn);
+
+        amountFeiOut = adjustedAmountIn
+            .mul(Constants.BASIS_POINTS_GRANULARITY - mintFeeBasisPoints)
+            .div(Constants.BASIS_POINTS_GRANULARITY)
+            .asUint256();
+    }
+
+    /// @notice helper function to get redeem amount out based on current market prices
+    /// @dev will revert if price is outside of bounds and bounded PSM is being used
+    function _getRedeemAmountOut(uint256 amountFeiIn)
+        internal
+        view
+        virtual
+        returns (uint256 amountTokenOut)
+    {
+        Decimal.D256 memory price = readOracle();
+        _validatePriceRange(price);
+
+        /// get amount of dollars being provided
+        Decimal.D256 memory adjustedAmountIn = Decimal.from(
+            (amountFeiIn *
+                (Constants.BASIS_POINTS_GRANULARITY - redeemFeeBasisPoints)) /
+                Constants.BASIS_POINTS_GRANULARITY
+        );
+
+        /// now turn the dollars into the underlying token amounts
+        /// dollars / price = how much token to pay out
+        amountTokenOut = adjustedAmountIn.div(price).asUint256();
+    }
+
+    /// @notice transfer ERC20 token to the recipient from the PCV Deposit
+    /// @param to recipient address
+    /// @param amount number of tokens sent to recipient
+    function _transfer(address to, uint256 amount) internal {
+        pcvDeposit.withdraw(to, amount);
+    }
+
+    /// @notice transfer assets from user to the PCV Deposit
+    /// Does not call deposit to save gas and make mints cheaper
+    /// @param from sending address
+    /// @param amount number of tokens sent to PCV Deposit
+    function _transferFrom(
+        address from,
+        address,
```

Comment:
Instead of hardcoding in the pcvDeposit here and ignoring the second parameter (???), it'd be much better to use the pcv deposit address in the function that *calls* this - ie, inside of mint()

Reply:
absolutely. This is weird as-is

Reply:
fixed